docs/documentation/coding-style.md

```diff
@@ -0,0 +1,161 @@
+# ILIAS Coding Standard
+This coding standard is for the [ILIAS project](https://github.com/ILIAS-eLearning/ILIAS)
+
+
+**Table of Contents**
+* [PSR-2](#psr-2)
+* [Additions](#additions)
+  * [Indention](#indention)
+  * [Exceptions](#exceptions)
+  * [Abstract](#abstract)
+  * [Interfaces](#interfaces)
+  * [Operators](#operators)
+  * [Type cast](#type-cast)
+  * [Return Type Declaration](#return-type-declaration)
+  * [Strings](#strings)
+  * [Comparison](#comparison)
+  * [Comments](#comments)
+  * [Early return](#early-return)
+  * [Native Types](#native-types)
+  * [Conditions](#conditions)
+  * [Arrays](#arrays)
+
+# PSR-2 
+
+The ILIAS coding standard is aligned to the widely used and established
+[PSR-2 standard](https://www.php-fig.org/psr/psr-2/)
+of the [PHP Interop Group (PHP-FIG)](https://www.php-fig.org/).
+
+_Info: Please be aware of the [additions](#additions)
+defined for the ILIAS project that MUST be used on top of PSR-2_
+
+# Additions
+
+This document is an addition to PSR-2 Standard used in the ILIAS project.
+This section was inspired by the
+[Doctrine Coding Standards](https://github.com/doctrine/coding-standard).
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Indention
+
+Divergent from the PSR-2 indentation, the code
+MUST be indent by tabs and NOT spaces.
+
+## Exceptions
+
+Abstract exception class names and exception interface names SHOULD be
+suffixed with `Exception`.
+
+## Abstract
+
+Abstract classes SHOULD NOT be prefixed or suffixed with Abstract.
+
+## Interfaces
+
+Interfaces SHOULD NOT be prefixed or suffixed with Interface.
+
+## Operators
+
+Operators MUST have a space before and after the operator.
+
+Mathematical Operators:
+
+```php
+$result = 1 + $value;
+```
+
+Comparsion Operators:
+
+```php
+if ($size > 0) {
+    // ...
+}
+```
+
+Concatenation Operator:
+
+```php
+$test = 'hello ' . $name . '!';
+```
+
+## Type cast
+
+Spaces MUST be added after a type cast.
+
+```php
+$foo = (int) '12345';
+```
+
+## Return Type Declaration
+
+Spaces MUST be added around a colon in return type declaration.
+
+```php
+function () : void {}
+```
+
+## Strings
+
+Strings MUST use single quotes(apostrophes) for enclosing tags.
+Double quotes MUST NOT be used for enclosing tags.
```

Comment:
Because I shouldn't need to look for things that are not strings inside strings. :stuck_out_tongue_closed_eyes: 

I would like to defend my argument with the argument, that the reviewer always needs to interpret the values during the review process.
The reviewer must be able to interpret the double quotes during the review process e.g.
`"\n"`, `"Hello $world"`.
So as reviewer I need to have to look exactly what the string contains if using double quote, because these are also potential bugs, if wrongly interpreted by the user. This misinterpretation is a general danger but I think interpreting the double quotes strings is a more potential for a misinterpretation of an string, while with single quotes the reviewer can be sure this is just a simple string.

I am aware that there can be some corner cases, like HTTP message (especially the interpretation of  the CRLF characters `\r\n`), but I also think that, we can avoid potential bugs that can slip through the review process here. IMO double quotes should be an exception of this rule and only be used in those corner cases and are always in the hand of the maintainer/reviewer.

I would love to here additional opinions on this matter. In every project this leads to discussion and I really think, that we **need** this discussions.